internal/core/textlayout/sharedparley.rs

```diff
@@ -0,0 +1,138 @@
+    pub text_overflow: TextOverflow,
+}
+
+impl Default for LayoutOptions {
+    fn default() -> Self {
+        Self {
+            max_width: None,
+            max_height: None,
+            horizontal_align: TextHorizontalAlignment::Left,
+            vertical_align: TextVerticalAlignment::Top,
+            stroke: None,
+            selection: None,
+            font_request: None,
+            text_wrap: TextWrap::WordWrap,
+            text_overflow: TextOverflow::Clip,
+        }
+    }
+}
+
+pub fn layout(text: &str, scale_factor: f32, options: LayoutOptions) -> Layout {
+    let mut font_context = font_context();
+    let mut layout_context = LAYOUT_CONTEXT.clone();
+
+    let mut builder = layout_context.ranged_builder(&mut font_context, text, scale_factor, true);
+    if let Some(ref font_request) = options.font_request {
+        if let Some(family) = &font_request.family {
+            builder.push_default(parley::StyleProperty::FontStack(
+                parley::style::FontStack::Single(parley::style::FontFamily::Named(
+                    family.as_str().into(),
+                )),
+            ));
+        }
+        if let Some(weight) = font_request.weight {
+            builder.push_default(parley::StyleProperty::FontWeight(
+                parley::style::FontWeight::new(weight as f32),
+            ));
+        }
+        if let Some(letter_spacing) = font_request.letter_spacing {
+            builder.push_default(parley::StyleProperty::LetterSpacing(letter_spacing.get()));
+        }
+        builder.push_default(parley::StyleProperty::FontStyle(if font_request.italic {
+            parley::style::FontStyle::Italic
+        } else {
+            parley::style::FontStyle::Normal
+        }));
+    }
+    let pixel_size = options
+        .font_request
+        .and_then(|font_request| font_request.pixel_size)
+        .unwrap_or(DEFAULT_FONT_SIZE);
+    builder.push_default(parley::StyleProperty::FontSize(pixel_size.get()));
+    builder.push_default(parley::StyleProperty::WordBreak(match options.text_wrap {
+        TextWrap::NoWrap => parley::style::WordBreakStrength::KeepAll,
+        TextWrap::WordWrap => parley::style::WordBreakStrength::Normal,
+        TextWrap::CharWrap => parley::style::WordBreakStrength::BreakAll,
+    }));
+    if options.text_overflow == TextOverflow::Elide {
+        todo!();
+    }
+
+    builder.push_default(parley::StyleProperty::Brush(Brush {
+        stroke: options.stroke,
+        ..Default::default()
+    }));
+    if let Some(selection) = options.selection {
+        builder.push(
+            parley::StyleProperty::Brush(Brush { stroke: options.stroke, is_selected: true }),
+            selection,
+        );
+    }
+
+    let mut layout: parley::Layout<Brush> = builder.build(text);
+    layout.break_all_lines(options.max_width.map(|max_width| max_width.get() * scale_factor));
+    layout.align(
+        options.max_width.map(|max_width| max_width.get() * scale_factor),
```

Comment:
The repeated multiplications of `max_width` make me wonder if it should perhaps be done once at the top of the function. Or, alternatively, the function could operate solely on `PhysicalLength` throughout (but that requires changing FontRequest?)